src/Autofac.Integration.AspNetCore.Multitenant/MultitenantRequestServicesMiddleware.cs

```diff
@@ -82,5 +115,29 @@ public async Task Invoke(HttpContext context)
                 context.Features.Set(existingFeature);
             }
         }
+
+        private async Task<IServiceScopeFactory> GetServiceScopeFactoryForTenantLifetimeScope(CancellationToken httpContextRequestAborted)
```

Comment:
Doesn't Autofac already handle locking around resolving singletons/instance-per-lifetime-scope? Are we putting a perf bottleneck in here that doesn't need to be here?

Reply:
Sadly we have to somehow manually handle that specific case here. I double checked, the implementation-factory is called several times for the same tenant, causing an `InstancePerDependency()` behavior even though I used `.SingleInstance()`. 

That's why I have added that caching mechanism here. Should the delegate really just be called once?

Reply:
I agree that the solution is not good in any case. I think I will leave it work today and get back to it again with some fresh thoughts - also trying to implement the specification as you have suggested. I think that would help.

Reply:
I removed the manual locking and storing of cached instances of `IServiceProviderFactory` for each tenant. Sadly, the `InstancePerTenant` does not act like an instance-per-tenant but I have no clue why yet. I will try to add the specification tests now and see how that goes. Maybe it will help to find the problem.